Support for New Architecture - iOS & Android (Fabric)

markrickert opened the pull request:
## Description

Fixes #196, #811, & #886

This PR builds on @j-piasecki's work in #550 to enable Fabric/New Architecture. That PR is fairly old and was created with `react-native@0.70.4` (lots of performance issues) and `FlashList@1.2.1`.

### Here's what I did to bring things back up to date so that people can start testing this with the latest version of FlashList:

1. Got @j-piasecki's `fabricfixture` app building properly.
2. Rebased this fabric branch with `@shopify/flash-list@master` to bring the library up to the latest version (`1.6.1`)
3. Upgraded the `fabricfixture` app from `react-native@0.70.4` to `0.72.5` ensuring that both iOS and android fixture apps built and displayed the `FlashList` component with the new architecture turned on.

### Things I did NOT do:

- I did not touch or edit ANY of the fabric code that @j-piasecki wrote. All credit goes to him.

### Things I think we still need to do:

- [ ] Investigate using [react-native-test-app](https://github.com/microsoft/react-native-test-app) so we don't have to have two fixture apps in the repo anymore (one for new architecture and another for old architecture). This will get rid of a lot of duplicated code as suggested by @fortmarek [here](https://github.com/Shopify/flash-list/pull/550#discussion_r948762557)
- [ ] Ensure that nothing broke in the non-fabric renderer.
- [ ] Test the he🏒🏒 out of this.

## Reviewers’ hat-rack :tophat:

- [ ] There is a new fixture app in `./fabricfixture` that can be used to test this functionatlity.

## Screenshots or videos (if needed)

These videos are from the app built by the react native app in `./fabricfixture`. Get ready to have a seizure, lol.

| iOS | Android |
|--------|--------|
| ![4361F280-4566-4146-9C74-5EF53CDABF67-76478-0001A1F3D40DDEDD](https://github.com/Shopify/flash-list/assets/139261/a29dc019-b3d7-4001-b516-8f9d39cd4b1c) | ![6E27994A-9439-4D04-95F5-B3531068FE01-5137-00028CE649BCEF28](https://github.com/Shopify/flash-list/assets/139261/313efff3-437e-4f48-8b4f-5f0c47ec29eb) |

## Checklist

- [x] I have added a changelog entry following the [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) guidelines.

markrickert commented:
I have signed the CLA!

RichardLindhout commented:
LGTM!

RichardLindhout commented:
I did not test it, I just was hoping FlashList would get Fabric support :). I'm not experienced with Fabric so I'm not sure.

naqvitalha commented:
Apologies for the delay, I'll try this out soon. Thanks a lot

WoLewicki commented on ios/Sources/AutoLayoutViewComponentView.mm:
```diff
@@ -0,0 +1,80 @@
+#ifdef RN_FABRIC_ENABLED
```
Change all of the `RN_FABRIC_ENABLED` to  `RCT_NEW_ARCH_ENABLED`. Otherwise it won't work correctly since we don't add such flag on new arch.

itsmuntadhar replied:
seems like an easy change, anyway can I contribute?

afaik I'd have to clone the branch then merge my edit to it for it be available in this PR, is this correct or is there a better/faster way?

itsmuntadhar replied:
https://github.com/markrickert/flash-list/pull/1 @markrickert

markrickert replied:
@WoLewicki I made this change to `RCT_NEW_ARCH_ENABLED`

chakravarthi-bb commented:
Can we release this??

evelant commented:
This is the last dependency our app preventing it from working with the new arch. Hopefully this can get released soon!

evelant commented:
@markrickert I tried your fork in my app and it seems to work perfect with the new architecture enabled!

markrickert commented:
I rebased my work here on `main`.

WoLewicki commented:
@markrickert could you maybe bump the version of `fabric` app to RN 0.73 to see if it works?

markrickert commented:
> @markrickert could you maybe bump the version of `fabric` app to RN 0.73 to see if it works?

That's a great idea. I'll work on that here in the next few days

BLOCKMATERIAL commented:
up please very important

JairajJangleSLIOT commented:
Can we get this lint check passed to unblock the PR from being approved?

markrickert commented:
This PR should be ready to go. I've updated the PR based on suggestions as well as ensured that both the `fixture` app and `fabricfixture` app build and run on the latest Mac OS (14.3) and the latest Xcode (15.2)

CI is finally passing all green ✅ and I rebased on `main` and force-pushed (sorry to anyone using my branch commit hashes directly in their project).

The `fabricfixture` app has been updated to react-native `0.72.10` for now.

WoLewicki commented on RNFlashList.podspec:
```diff
@@ -2,21 +2,46 @@ require 'json'
-  # Dependencies
-  s.dependency 'React-Core'
+    # Don't install the dependencies when we run `pod install` in the old architecture.
+    if fabric_enabled
```
Yeah, we should drop it.

patrikduksin commented:
bump! would be happy to see that in the next release

JairajJangleSLIOT commented on ios/Sources/AutoLayoutView.swift:
```diff
@@ -85,15 +99,21 @@ import UIKit
+            .compactMap { subview -> CellContainerComponentView? in
+                if let cellContainer = subview as? CellContainerComponentView {
                     return cellContainer
+                } else if subview is AutoLayoutView {
```
@WoLewicki @markrickert just wanted to check: apart from this comment are there any other issues with this PR which is blocking the merge?

WoLewicki replied:
Nope, it looks ok

dkjepben commented on ios/Sources/AutoLayoutViewComponentView.mm:
```diff
@@ -0,0 +1,80 @@
+#import <react/renderer/components/rnflashlist/RCTComponentViewHelpers.h>
+
+#import "RCTFabricComponentsPlugins.h"
+#import <RNFlashList/RNFlashList-Swift.h>
```
In our project we had to change this to:
#import <RNFlashList-Swift.h>

Or else xcode cannot find it.

Xcode 15.3
react-native 0.73.4

anton-patrushev replied:
```suggestion
#import <RNFlashList-Swift.h>
```

anton-patrushev commented:
Any updates?
Does anyone have time to chat about the required help on this PR?

cc: @markrickert @WoLewicki @naqvitalha

WoLewicki reviewed (approved):
LGTM!

markrickert commented:
@WoLewicki I still can't seem to get the iOS CI build working. I think i'm close, but i got frustrated and gave up yesterday, haha.

markrickert commented:
Weird. They weren't passing yesterday afternoon. 🤷 :shipit:

anton-patrushev commented:
@gvarandas @naqvitalha 

Can you take a look? We need this PR to be released asap 🙏

anton-patrushev reviewed (approved):
When can we get this in?

cc: @markrickert @naqvitalha

markrickert commented:
Rebased on `main` and fixed the merge conflicts.

GeorgiIvanovPXL commented:
Is this going to be merged soon? 🤔

jamonholmgren reviewed (approved):
Nice work on this, @markrickert! (And @j-piasecki too!)

anton-patrushev reviewed (approved):
Great job!

qq: is it compatible with latest 74 RN release?

ChronoByteCosmonaut commented:
What’s the status on this one?

gorbypark commented:
> Great job!
> 
> qq: is it compatible with latest 74 RN release?

Just a very quick test, but I spun up a new Expo 51 project (which is RN 0.74.0) and this pr seems to work great on both iOS/Android with new arch bridgless mode enabled.

```bash
# create a new project with SDK51 (as of April 28, 2024 this is the beta version)
yarn create expo-app --template default@beta

# this just removes the examples in the app folder
yarn reset-project

# required to enable new architecture
npx expo install expo-build-properties

# use flashlist from this PR
yarn add https://github.com/markrickert/flash-list.git\#mrickert/fabric

# generate android and ios folders
npx expo prebuild --clean
```

`app.json` (to enables new architecture)
```json
    "plugins": [
      "expo-router",
      [
        "expo-build-properties",
        {
          "ios": {
            "newArchEnabled": true
          },
          "android": {
            "newArchEnabled": true
          }
        }
      ]
    ],
 ```
    
`/app/index.tsx`
   ```jsx
import React from "react";
import { Text } from "react-native";
import { FlashList } from "@shopify/flash-list";

const DATA = [
  {
    title: "First Item",
  },
  {
    title: "Second Item",
  },
];

export default function App() {
return (
    <FlashList
      data={DATA}
      renderItem={({ item }) => <Text>{item.title}</Text>}
      estimatedItemSize={200}
    />
  );
}
```
Build and run app
```bash
yarn ios
yarn android
```

A very simple usage of flashlist (copied directly from the flashlist doc site) but both platforms are working as expected.

almoatamed commented:
this is very important, is this going to be merged?

WoLewicki commented:
I am not a maintainer of this library unfortunately.

saxenanickk commented:
Anyone maintaining this library?

saxenanickk commented:
@naqvitalha @jenshenny Anyone maintaining this library?

jfmoe commented:
Any progress?

j-piasecki commented on RNFlashList.podspec:
```diff
@@ -9,14 +9,19 @@ Pod::Spec.new do |s|
 
-  # Dependencies
-  s.dependency 'React-Core'
+  # Use install_modules_dependencies helper to install the dependencies if React Native version >=0.71.0.
```
@naqvitalha The `RCT_NEW_ARCH_ENABLED` isn't actually set anywhere.

```suggestion
  if fabric_enabled
    s.pod_target_xcconfig = {
      'OTHER_SWIFT_FLAGS' => '-D RCT_NEW_ARCH_ENABLED',
    }
  end

  # Use install_modules_dependencies helper to install the dependencies if React Native version >=0.71.0.
```

And then it would also be required to check if the flag is set at the top level (cannot add suggestion there 😞):
```ruby
fabric_enabled = ENV['RCT_NEW_ARCH_ENABLED']
```

markrickert commented:
Closing because #550 was merged.
